PerspectiveRailroad/PerspectiveRailroad.js

```diff
@@ -0,0 +1,302 @@
+    }
+
+    this.quad = quad;
+    this.start_frame = start_frame;
+    this.duration_frames = duration_frames;
+  }
+
+  compute_polygon(frame) {
+    // We haven't started animating, so don't render anything
+    if (frame < this.start_frame) {
+      return undefined;
+    }
+
+    // We've finished the animation, so just render the full quad
+    if (frame >= this.start_frame + this.duration_frames) {
+      return this.quad;
+    }
+
+    // In between, expand the tie from the left edge
+    const t = (frame - this.start_frame) / this.duration_frames;
+    const [a, b, c, d] = this.quad.vertices;
+    const bottom_right = Point.lerp(a, b, t);
+    const top_right = Point.lerp(d, c, t);
+
+    return new PolygonPrimitive([a, bottom_right, top_right, d]);
+  }
+}
+
+const RAILS_START_FRAME = 20;
+const RAILS_DURATION = 200;
+const TIE_DURATION = 30;
+function make_ties() {
+  const tie_bottom_left = Point.point(50, HEIGHT);
+  const tie_bottom_right = Point.point(475, HEIGHT);
+
+  const TIE_SPACING = 100;
+  const tie_bottoms = even_spaced_rectangles(
+    tie_bottom_left,
+    tie_bottom_right,
+    VP_RAILS,
+    TIE_SPACING
+  );
+  const tie_prims = railroad_ties(tie_bottoms, 0.5);
+
+  const animated_ties = tie_prims.map((quad, i) => {
+    const y = quad.vertices[0].y;
+    const t_start = (y - HEIGHT) / (VP_RAILS.y - HEIGHT);
+    const start_frame = t_start * RAILS_DURATION;
+
+    return new AnimatedTie(quad, start_frame, TIE_DURATION);
+  });
+  return animated_ties;
+}
+
+class AnimatedRails {
+  constructor(triangles, start_frame, duration_frames) {
+    triangles.forEach((x) => {
+      if (x.vertices.length !== 3) {
+        throw new Error("Each rail primitive must be a triangle!");
+      }
+    });
+    this.triangles = triangles;
+    this.start_frame = start_frame;
+    this.duration_frames = duration_frames;
+  }
+
+  compute_polygons(frame) {
+    if (frame < this.start_frame) {
+      return [];
+    }
+
+    if (frame >= this.start_frame + this.duration_frames) {
+      return this.triangles;
+    }
+
+    // In between, lerp points towards the vanishing point. This will
+    // produce a _quad_ rather than a triangle for the in-between frames.
+    const t = (frame - this.start_frame) / this.duration_frames;
+    return this.triangles.map((tri) => {
+      const [a, b, vp] = tri.vertices;
+
+      const top_left = Point.lerp(a, vp, t);
+      const top_right = Point.lerp(b, vp, t);
+      return new PolygonPrimitive([a, b, top_right, top_left]);
+    });
+  }
+}
+
+function make_rails() {
+  const rail_prims = compute_rails();
+
+  return new AnimatedRails(rail_prims, RAILS_START_FRAME, RAILS_DURATION);
+}
+
+const BACKGROUND = make_background();
+const ANIMATED_TIES = make_ties();
+const ANIMATED_RAILS = make_rails();
+
+export const sketch = (p) => {
+  p.setup = () => {
+    p.createCanvas(WIDTH, HEIGHT);
+
+    draw_primitive(p, BACKGROUND);
+  };
+
+  p.draw = () => {
```

Comment:
After the railroad finishes animating, there's no need to render further frames. Maybe add a check at the end of the loop to see if all the animation primitives are past `start_frame + duration`? Maybe add `get is_finished()` to both types of animation primitives